File: src/platform.ts
/**
 * Cross-platform detection and abstraction for ccbox.
 *
 * Provides a unified interface for platform-specific behaviors:
 * - Host platform detection (windows-wsl, windows-native, macos, linux)
 * - Docker socket paths
 * - Clipboard commands
 * - FUSE requirement detection
 *
 * Dependency direction:
 *   This module has minimal internal dependencies (near-leaf module).
 *   It may be imported by: docker-runtime.ts, paths.ts, clipboard.ts, voice.ts
 *   It should NOT import from: cli, generator, docker-runtime
 */

import { readFileSync } from "node:fs";
import { platform } from "node:os";
import { env } from "node:process";

/** Supported host platform types. */
export type HostPlatform = "windows-wsl" | "windows-native" | "macos" | "linux";

/** Cached platform detection result. */
let cachedPlatform: HostPlatform | null = null;

/**
 * Detect the host platform.
 *
 * Detection order:
 * 1. Windows native (process.platform === "win32")
 * 2. WSL (Linux kernel with "microsoft" in /proc/version)
 * 3. macOS (process.platform === "darwin")
 * 4. Linux (fallback)
 *
 * Result is cached for performance.
 */
export function detectHostPlatform(): HostPlatform {
  if (cachedPlatform !== null) {
    return cachedPlatform;
  }

  const os = platform();

  if (os === "win32") {
    cachedPlatform = "windows-native";
    return cachedPlatform;
  }

  if (os === "darwin") {
    cachedPlatform = "macos";
    return cachedPlatform;
  }

  // Linux — check if WSL
  if (os === "linux") {
    try {
      const procVersion = readFileSync("/proc/version", "utf-8");
      if (procVersion.toLowerCase().includes("microsoft")) {
        cachedPlatform = "windows-wsl";
        return cachedPlatform;
      }
    } catch {
      // Can't read /proc/version — not WSL
    }

    // Fallback: check WSL env vars
    if (env.WSL_DISTRO_NAME ?? env.WSLENV) {
      cachedPlatform = "windows-wsl";
      return cachedPlatform;
    }
  }

  cachedPlatform = "linux";
  return cachedPlatform;
}

/**
 * Whether the host platform requires FUSE for path translation.
 *
 * FUSE is needed on Windows (WSL and native) where host paths differ
 * from container POSIX paths. macOS and Linux don't need path translation
 * since paths are already POSIX-compatible.
 */
export function needsFuse(): boolean {
  const p = detectHostPlatform();
  return p === "windows-wsl" || p === "windows-native";
}

/**
 * Whether the host platform needs --privileged for FUSE /dev/fuse access.
 *
 * Windows Docker Desktop requires --privileged because /dev/fuse is not
 * directly passable via --device. Linux/macOS use --device /dev/fuse.
 */
export function needsPrivilegedForFuse(): boolean {
  const p = detectHostPlatform();
  return p === "windows-native";
}

/**
 * Get the Docker socket path for the current platform.
 */
export function getDockerSocketPath(): string {
  const p = detectHostPlatform();
  switch (p) {
    case "windows-native":
      return "//./pipe/docker_engine";
    case "windows-wsl":
    case "linux":
    case "macos":
      return "/var/run/docker.sock";
  }
}

/**
 * Get the clipboard read command for the current platform.
 *
 * Returns the command and arguments to read image data from clipboard.
 * Returns null if no clipboard command is available.
 */
export function getClipboardImageCommand(): { cmd: string; args: string[] } | null {
  const p = detectHostPlatform();
  switch (p) {
    case "windows-native":
    case "windows-wsl":
      return {
        cmd: "powershell.exe",
        args: ["-NoProfile", "-Command", "Get-Clipboard -Format Image | ForEach-Object { $_.Save([Console]::OpenStandardOutput(), [System.Drawing.Imaging.ImageFormat]::Png) }"],
      };
    case "macos":
      return {
        cmd: "osascript",
        args: ["-e", "set png to (the clipboard as «class PNGf»)"],
      };
    case "linux": {
      // Prefer wl-paste for Wayland, fall back to xclip for X11
      if (env.WAYLAND_DISPLAY) {
        return { cmd: "wl-paste", args: ["--type", "image/png"] };
      }
      if (env.DISPLAY) {
        return { cmd: "xclip", args: ["-selection", "clipboard", "-t", "image/png", "-o"] };
      }
      return null;
    }
  }
}

/**
 * Get the clipboard text read command for the current platform.
 */
export function getClipboardTextCommand(): { cmd: string; args: string[] } | null {
  const p = detectHostPlatform();
  switch (p) {
    case "windows-native":
    case "windows-wsl":
      return { cmd: "powershell.exe", args: ["-NoProfile", "-Command", "Get-Clipboard"] };
    case "macos":
      return { cmd: "pbpaste", args: [] };
    case "linux": {
      if (env.WAYLAND_DISPLAY) {
        return { cmd: "wl-paste", args: [] };
      }
      if (env.DISPLAY) {
        return { cmd: "xclip", args: ["-selection", "clipboard", "-o"] };
      }
      return null;
    }
  }
}

/**
 * Whether path translation is needed between host and container.
 *
 * On Windows (both WSL and native), host paths use different formats
 * than container POSIX paths. macOS and Linux use POSIX paths natively.
 */
export function needsPathTranslation(): boolean {
  const p = detectHostPlatform();
  return p === "windows-wsl" || p === "windows-native";
}

/**
 * Get human-readable host OS name for display.
 */
export function getHostOSName(): string {
  const p = detectHostPlatform();
  switch (p) {
    case "windows-native":
    case "windows-wsl":
      return "Windows";
    case "macos":
      return "macOS";
    case "linux":
      return "Linux";
  }
}

/**
 * Check if a command exists on the host system.
 */
export function commandExists(command: string): boolean {
  try {
    const { execSync } = require("node:child_process");
    execSync(`which ${command}`, { stdio: "ignore" });
    return true;
  } catch {
    return false;
  }
}

/** Reset cached platform (for testing). */
export function _resetPlatformCache(): void {
  cachedPlatform = null;
}

File: internal/platform/platform.go
// Package platform provides host platform detection and platform-specific
// configuration for Docker integration.
package platform

import (
	"os"
	"os/exec"
	"runtime"
	"strings"
	"sync"
)

// HostPlatform represents the detected host operating system environment.
type HostPlatform string

const (
	// WindowsWSL is Windows Subsystem for Linux (Docker Desktop via WSL backend).
	WindowsWSL HostPlatform = "windows-wsl"
	// WindowsNative is native Windows (Docker Desktop, Git Bash, PowerShell).
	WindowsNative HostPlatform = "windows-native"
	// MacOS is macOS (Docker Desktop).
	MacOS HostPlatform = "macos"
	// Linux is native Linux (Docker Engine).
	Linux HostPlatform = "linux"
)

var (
	detectedPlatform HostPlatform
	detectOnce       sync.Once
)

// DetectHost returns the current host platform, caching the result.
//
// Detection logic:
//   - runtime.GOOS == "windows" -> WindowsNative
//   - runtime.GOOS == "darwin"  -> MacOS
//   - runtime.GOOS == "linux" + /proc/version contains "microsoft" -> WindowsWSL
//   - runtime.GOOS == "linux" (fallback) -> Linux
func DetectHost() HostPlatform {
	detectOnce.Do(func() {
		switch runtime.GOOS {
		case "windows":
			detectedPlatform = WindowsNative
		case "darwin":
			detectedPlatform = MacOS
		case "linux":
			if isWSL() {
				detectedPlatform = WindowsWSL
			} else {
				detectedPlatform = Linux
			}
		default:
			// Unknown OS, assume Linux-like behavior
			detectedPlatform = Linux
		}
	})
	return detectedPlatform
}

// isWSL checks if running inside WSL by reading /proc/version.
func isWSL() bool {
	// Check /proc/version for "microsoft" (covers WSL1 and WSL2)
	data, err := os.ReadFile("/proc/version")
	if err == nil && strings.Contains(strings.ToLower(string(data)), "microsoft") {
		return true
	}
	// Fallback: check WSL-specific environment variables
	if os.Getenv("WSL_DISTRO_NAME") != "" || os.Getenv("WSLENV") != "" {
		return true
	}
	return false
}

// NeedsFuse returns whether the platform requires FUSE for path translation.
// Windows (both native and WSL) needs FUSE because Claude Code stores absolute
// host paths in config files that differ from container paths.
func NeedsFuse() bool {
	p := DetectHost()
	return p == WindowsNative || p == WindowsWSL
}

// NeedsPrivilegedForFuse returns whether the platform needs --privileged
// for FUSE to work. Only WindowsNative needs this because Docker Desktop
// on Windows does not support --device=/dev/fuse.
func NeedsPrivilegedForFuse() bool {
	return DetectHost() == WindowsNative
}

// DockerSocket returns the Docker socket path for the current platform.
// Windows native uses a named pipe; all others use the Unix socket.
func DockerSocket() string {
	if DetectHost() == WindowsNative {
		return "//./pipe/docker_engine"
	}
	return "/var/run/docker.sock"
}

// ClipboardImageCmd returns the command and arguments for reading an image
// from the system clipboard, or nil if not supported on the current platform.
func ClipboardImageCmd() []string {
	switch DetectHost() {
	case WindowsNative:
		return []string{"powershell", "-NoProfile", "-Command",
			"Add-Type -AssemblyName System.Windows.Forms; " +
				"$img = [System.Windows.Forms.Clipboard]::GetImage(); " +
				"if ($img) { $img.Save([Console]::OpenStandardOutput(), [System.Drawing.Imaging.ImageFormat]::Png) }"}
	case MacOS:
		return []string{"osascript", "-e",
			"set png to (the clipboard as «class PNGf»)",
			"-e", "set f to open for access POSIX file \"/dev/stdout\" with write permission",
			"-e", "write png to f",
			"-e", "close access f"}
	case Linux, WindowsWSL:
		if os.Getenv("WAYLAND_DISPLAY") != "" {
			return []string{"wl-paste", "--type", "image/png"}
		}
		return []string{"xclip", "-selection", "clipboard", "-t", "image/png", "-o"}
	default:
		return nil
	}
}

// ClipboardTextCmd returns the command and arguments for reading text
// from the system clipboard, or nil if not supported.
func ClipboardTextCmd() []string {
	switch DetectHost() {
	case WindowsNative:
		return []string{"powershell", "-NoProfile", "-Command", "Get-Clipboard"}
	case MacOS:
		return []string{"pbpaste"}
	case Linux, WindowsWSL:
		if os.Getenv("WAYLAND_DISPLAY") != "" {
			return []string{"wl-paste"}
		}
		return []string{"xclip", "-selection", "clipboard", "-o"}
	default:
		return nil
	}
}

// NeedsPathTranslation returns whether Docker volume mount paths need
// translation on this platform. Windows paths (D:\x) must become D:/x,
// and WSL paths (/mnt/d/x) must become /d/x.
func NeedsPathTranslation() bool {
	p := DetectHost()
	return p == WindowsNative || p == WindowsWSL
}

// HostOSName returns a human-readable OS name string.
func HostOSName() string {
	switch DetectHost() {
	case WindowsNative:
		return "Windows"
	case WindowsWSL:
		return "Windows (WSL)"
	case MacOS:
		return "macOS"
	case Linux:
		return "Linux"
	default:
		return "Unknown"
	}
}

// CommandExists checks whether a command is available in the system PATH.
func CommandExists(name string) bool {
	_, err := exec.LookPath(name)
	return err == nil
}
